### PR TITLE
feat(server): add /watermark and /watermark/batch endpoints with sidecar support (#313)

### LINKED ISSUE #313: Feature Proposal: /watermark API endpoint for native SynthID text generation
Hey @guillaumemeyer, following up on our X thread! 

Proposing the design to implement a native `/watermark` endpoint. This gives us a built-in way to generate watermarked text for benchmarks without relying on the heavy `MarkLLM` harness. Per your note, I'm keeping this strictly scoped to the API for now.

### 1. API Endpoint Design
Following the routing pattern in `server.py`, we can expose `POST /watermark`. I noticed your `/clean` and `/inspect` endpoints use base64 encoding for the `"file"` payload. For this generation endpoint, we can accept raw strings for simplicity, or stick to base64 if you prefer strict consistency across the API. I've drafted it with raw strings below.

**Request:**
```json
{
  "text": "The raw text or prompt to be watermarked.",
  "keys": [118, 504, 421, 521], 
  "options": {
    "scheme": "synthid"
  }
}
```

**Response:**
```json
{
  "ok": true,
  "kind": "text",
  "watermarked_text": "...",
  "report": {
    "scheme_used": "synthid"
  }
}
```

### 2. Implementation Approach
* **Respecting the `stdlib-only` core:** Since `wr-core` is strictly Python 3.10+ stdlib-only, adding a logits processor directly would bloat the core requirements. Instead, we can design this as an optional backend/sidecar (similar to your `wr-synthid-score` setup) to keep the core service dependency-free while isolating the generation logic.
* **Batching Integration:** Wire this up to a `POST /watermark/batch` route to perfectly match your existing `/clean/batch` and `/inspect/batch` patterns.

Let me know if this payload structure and sidecar approach aligns with what you had in mind, and I can start putting together the PR!

### PR DESCRIPTION
**Pull request by kilopal**
Closes #313
Summary
Adds POST /watermark and POST /watermark/batch to wr-core along with a dedicated wr-synthid-text HTTP sidecar (port 8767) under the harness profile.
Following the maintainer's recommendation in #313, this replicates the established wr-synthid-score sidecar pattern for text. It lets wr-core generate native SynthID-watermarked text for benchmarks and round-trip verification over HTTP without bundling heavy ML dependencies (PyTorch, Transformers, MarkLLM) into the published core image.
Key Additions & Architecture
1. Core Service (server.py)
- Implements POST /watermark and POST /watermark/batch.
- Accepts raw text string or base64-encoded file payload.
- Enforces binary guards (looks_binary) to reject binary files from text endpoints.
- Batch isolation: each entry is caught in a per-file try/except; errors surface under that entry's "ok": false and never abort the rest of the batch.
- Deadline-bounded batch processing derived from the sidecar timeout, so a slow entry cannot hang the request.
- Full OpenAPI 3.0.3 spec with request bounds, additionalProperties: false, and documented 400/502/503 error responses; exposed via /capabilities (text_generators.synthid_http).
2. Text Sidecar (synthid_text_server.py)
- Lightweight stdlib ThreadingHTTPServer on port 8767 with optional bearer-key auth (WATERMARKS_SYNTHID_TEXT_API_KEY).
- Keeps the MarkLLM model resident under a thread lock (_MODEL_LOCK) with a bounded LRU cache to eliminate per-request model-loading overhead.
- Applies custom SynthID keys at model construction so watermarking matches the requested keys, not the config file defaults.
- Validates option types early (max_new_tokens, min_length, seed) and rejects non-UTF-8 uploads and byte caps with an immediate 400.
3. Client Dispatch (text_watermark.py)
- Fail-soft dispatch hierarchy: HTTP sidecar (WATERMARKS_SYNTHID_TEXT_URL) → local MarkLLM checkout (MARKLLM_DIR) → informative unconfigured error.
- Normalizes custom integer keys (int list, comma-separated string, or JSON array) and rejects booleans.
- SSRF protection: a custom _NoRedirect handler refuses 301/302/307/308 redirects, and is applied by default even if urllib.request.urlopen is monkeypatched.
- Timeout resolution (resolve_timeout) clamps values to [1s, 600s] with an env default, applied to both single and batch paths.
- Correct error classification: request-validation 4xx → 400, sidecar 5xx/unreachable/timeout → 502, unconfigured → 503, and sidecar hit with a bad/missing bearer token (401) → 502 (gateway config error).
4. Docker Compose (compose.yaml) & Config
- Registers wr-synthid-text under the harness profile using Dockerfile.markllm.
- Container hardening mirrors wr-synthid-score: read_only: true, unprivileged user 10001:10001, init: true, and /tmp tmpfs.
- Documented environment variables in .env.example and README.md.
- (Minor doc fix) Added the missing /detect/batch row to the README endpoints table.
Validation
- Lint & Format: python -m ruff check service tests (0 errors) and python -m ruff format --check service tests (clean across all 102 files).
- OpenAPI 3.0.3 Contract: spec serializes validly, uses the boolean exclusiveMinimum form, and preserves explicit 502/503 response descriptors.
- Tests: pytest tests/test_synthid_text_server.py tests/test_watermark_endpoint.py (45 passed) — covering health, bearer auth, single/batch generation, batch isolation, caps & byte limits, binary rejection, keys normalization/boolean rejection, non-UTF-8 uploads, timeout clamping, empty-vs-default keys cache isolation, and SSRF redirect refusal.
- Full suite: 1070 passed, 143 skipped (excluding a pre-existing, watermark-unrelated Windows hook-test flake).

### CODE REVIEW

**Comment by guillaumemeyer** [comment]
@kilopal, still some work is required.
CodeRabbit comments + Blockers/fix before merge

1. Local-backend timeout is defeated — text_watermark._watermark_local_markllm:

with concurrent.futures.ThreadPoolExecutor(max_workers=1) as executor:
    future = executor.submit(wm.generate_watermarked_text, text)
    watermarked = future.result(timeout=timeout)

future.result(timeout=...) raises TimeoutError, but the with ThreadPoolExecutor exit path calls shutdown(wait=True), which blocks until the running generation finishes anyway. So the timeout/deadline doesn't actually bound the wait for the local MarkLLM path — a stuck generation overruns both the single-request timeout and the batch deadline. Use a long-lived single worker with executor.shutdown(wait=False)/future.cancel() on timeout instead of the context manager.

2. HTTP status is derived by substring-matching error strings — server._watermark_error_status matches "sidecar authentication error", "unreachable", "timed out", etc. Reword any message in text_watermark.py and the status silently changes. Have the client return a stable error_code (e.g. auth | client_error | unreachable | timeout | unconfigured) and map that.

3. Options the backend reads but the validator forbids — parse_watermark_options rejects device, config, offline, yet the sidecar (_generate_watermarked_sample) and the local path both do opts.get("device"/"config"/"offline"). Through the API those are unreachable, so the sidecar can never be told to use a device, a different config, or offline mode. Allow them (with the same validation pattern) or delete the reads.

**Comment by kilopal** [comment]
> @kilopal, still some work is required. CodeRabbit comments + Blockers/fix before merge
> 
> 1. Local-backend timeout is defeated — text_watermark._watermark_local_markllm:
> 
> with concurrent.futures.ThreadPoolExecutor(max_workers=1) as executor: future = executor.submit(wm.generate_watermarked_text, text) watermarked = future.result(timeout=timeout)
> 
> future.result(timeout=...) raises TimeoutError, but the with ThreadPoolExecutor exit path calls shutdown(wait=True), which blocks until the running generation finishes anyway. So the timeout/deadline doesn't actually bound the wait for the local MarkLLM path — a stuck generation overruns both the single-request timeout and the batch deadline. Use a long-lived single worker with executor.shutdown(wait=False)/future.cancel() on timeout instead of the context manager.
> 
> 2. HTTP status is derived by substring-matching error strings — server._watermark_error_status matches "sidecar authentication error", "unreachable", "timed out", etc. Reword any message in text_watermark.py and the status silently changes. Have the client return a stable error_code (e.g. auth | client_error | unreachable | timeout | unconfigured) and map that.
> 3. Options the backend reads but the validator forbids — parse_watermark_options rejects device, config, offline, yet the sidecar (_generate_watermarked_sample) and the local path both do opts.get("device"/"config"/"offline"). Through the API those are unreachable, so the sidecar can never be told to use a device, a different config, or offline mode. Allow them (with the same validation pattern) or delete the reads.

**Comment by kilopal** [comment]
@guillaumemeyer Thanks for the detailed review! I've pushed updates addressing all 3 blockers:

1. **Timeout Deadlock**: Swapped the `ThreadPoolExecutor` context manager for a persistent module-level single-worker pool (`_LOCAL_GEN_EXECUTOR`). On timeout, we cancel the future and return immediately so it no longer blocks on `shutdown(wait=True)`.
2. **HTTP Status Mapping**: Replaced fragile substring matching with a deterministic `error_code` contract across all failure dictionaries (`"auth"`, `"timeout"`, `"unreachable"`, `"client_error"`, etc.). `_watermark_error_status()` now maps directly from `error_code` to `HTTPStatus`.
3. **Options Passthrough & Validation**: Added `device`, `config`, and `offline` to `ALLOWED_WATERMARK_OPTIONS`. They are strictly validated in `parse_watermark_options()` and documented in the OpenAPI request schema.

All 46 unit & integration tests and ruff lint/formatting checks pass cleanly. Let me know if you need anything else to get this merged!

**Review by poorvith-mp — APPROVED** [review]
Architecture and implementation look solid and strictly respect the repository's core constraints:

1. **Sidecar Isolation**: Heavy ML dependencies (MarkLLM, PyTorch, Transformers) remain strictly quarantined inside the wr-synthid-text sidecar under the harness profile on port 8767, preserving wr-core's zero-dependency stdlib model.
2. **Defensive Hardening**: Solid SSRF prevention via _NoRedirect, bounded batch deadlines, binary payload rejections (looks_binary), and accurate error status boundaries (502 for sidecar unreachable/401 auth issues, 503 unconfigured, 400 validation).
3. **Verification**: Checked out locally and verified:
   - pytest tests/test_synthid_text_server.py tests/test_watermark_endpoint.py: 46 passed.
   - 
uff check service tests: Clean across all files.

LGTM!
